transports/p2p/connmgr/connmanager.go

```diff
@@ -209,17 +232,25 @@ func (cm *ConnManager) handleFailedConn(c *ConnReq) {
 			cm.Connect(c)
 		})
 	} else if cm.cfg.GetNewAddress != nil {
-		cm.failedAttempts++
-		if cm.failedAttempts >= maxFailedAttempts {
-			cm.log.Debug().Msgf("Max failed connection attempts reached: [%d] "+
-				"-- retrying connection in: %v", maxFailedAttempts,
-				cm.cfg.RetryDuration)
-			time.AfterFunc(cm.cfg.RetryDuration, func() {
-				cm.NewConnReq()
-			})
+		if c.Addr != nil && c.Addr.String() != "" && cm.cfg.BanAddress != nil {
+			cm.incAddrConnAttempt(c.Addr.String())
+			if cm.getAddrConnAttempt(c.Addr.String()) >= maxFailedAttempts {
+				cm.cfg.BanAddress(c.Addr.String())
+				return
+			}
 		} else {
-			go cm.NewConnReq()
+			cm.globalFailedAttempts++
+			if cm.globalFailedAttempts >= maxFailedAttempts {
```

Comment:
Is this thread safe? I mean - what happend when multiple goroutines try to modify or check the `globalFailedAttempts`?

Reply:
You're right, it was implemented like that previously. I changed it to be thread safe